Proposal: Metrics_Specs

t03apt opened the pull request:
### Update:
See: https://github.com/MassTransit/MassTransit/pull/5420#issuecomment-2299957003

### Obsolete:
~~Currently, there are no tests for metrics.
This PR is just a draft showing how MetricCollector could be used to verify tags.~~

~~[MetricCollector ](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.diagnostics.metrics.testing.metriccollector-1?view=net-8.0) is very new. It was introduced in v9.0.0-preview.6.24353.1.~~

NooNameR commented on tests/MassTransit.Metrics.Tests/Metrics_Specs.cs:
```diff
@@ -0,0 +1,95 @@
+                {
+                    x.AddConsumer<MonitoredSubmitOrderConsumer>();
+
+                    x.UsingActiveMq((context, cfg) =>
```
not ActiveMq please, why not just InMemory? or otherwise rmq

t03apt replied:
I wanted to use the InMemory, but the logging methods were not triggered for some reason. In a few days, I can try it again. I am afk.

t03apt replied:
> I'd rename them to OpenTelemetry.Tests (something more generic)
I renamed the test project to MassTransit.Observability.Tests.

> not ActiveMq please, why not just InMemory? or otherwise rmq
`InMemory` transport didn't work for me. I switched to rmq.

NooNameR reviewed (commented):
I do not mind such tests, but I'd rename them to OpenTelemetry.Tests (something more generic).
Also, I suggest to add tests for custom metrics

NooNameR commented on tests/MassTransit.Metrics.Tests/Metrics_Specs.cs:
```diff
@@ -0,0 +1,95 @@
+        [Test]
+        public async Task Should_get_metrics_with_tags()
+        {
+            var collector = new MetricCollector<long>(null, InstrumentationOptions.MeterName, "messaging.masstransit.consume");
```
don't you need to dispose it?

NooNameR replied:
and another thing, you can retrieve metric name from `InstrumentationOptions` options

t03apt replied:
I’ll check it. If it is disposable then it should be most probably disposed.

t03apt replied:
> and another thing, you can retrieve metric name from `InstrumentationOptions` options

I wouldn’t rely on too many external classes for assert. I would rather use some kind of a snapshot testing using Verify.NUnit and a json verification file.

NooNameR replied:
I'm not sure what you mean, but MetricName for collector is defined in InstrumentationOptions, so instead of constant metric name I'd use one in options.

t03apt commented:
@NooNameR 
I refactored the test.
1. InMemory transport didn't work for me. I replaced ActiveMq with RabbitMq.
2. I initialized many MetricCollectors using reflection on InstrumentationOptions.
3. Change the assert logic to snapshot testing. See: Metrics_Specs.Should_get_metrics_with_tags.verified.txt
4. I realized that MetricCollector exists in .net8. I moved the test class to MassTransit.RabbitMqTransport.Tests project.

Observations:
1. Sometimes the test fails because the `messaging.masstransit.destination` attribute is inconsistent.
2. `messaging.masstransit.destination` is missing on many metrics. https://github.com/MassTransit/MassTransit/pull/5416
3. Custom tags can be added only in IReceiveObserver.PreReceive. Tags added in other methods are missing. I think this is fine and could be documented as expected behavior.
4. Custom tags added in ISendObserver are missing. I think this should be fixed.

NooNameR commented:
> * Custom tags can be added only in IReceiveObserver.PreReceive. Tags added in other methods are missing. I think this is fine and could be documented as expected behavior.
> * Custom tags added in ISendObserver are missing. I think this should be fixed.

you should be using filters for this not observers

NooNameR commented:
> * InMemory transport didn't work for me. I replaced ActiveMq with RabbitMq.

I'll take a look into this

NooNameR commented:
I have just created a test and everything works with InMemory: https://github.com/MassTransit/MassTransit/pull/5429
